src/circuit/gadget/ecc/chip/add.rs

```diff
@@ -0,0 +1,385 @@
+        region: &mut Region<'_, C::Base>,
+    ) -> Result<EccPoint<C>, Error> {
+        // Enable `q_add` selector
+        self.q_add.enable(region, offset)?;
+
+        // Copy point `a` into `x_p`, `y_p` columns
+        util::assign_and_constrain(region, || "x_p", self.x_p, offset, &a.x, &self.perm)?;
+        util::assign_and_constrain(region, || "y_p", self.y_p, offset, &a.y, &self.perm)?;
+
+        // Copy point `b` into `x_qr`, `y_qr` columns
+        util::assign_and_constrain(region, || "x_q", self.x_qr, offset, &b.x, &self.perm)?;
+        util::assign_and_constrain(region, || "y_q", self.y_qr, offset, &b.y, &self.perm)?;
+
+        let (x_p, y_p) = (a.x.value, a.y.value);
+        let (x_q, y_q) = (b.x.value, b.y.value);
+
+        // inv0(x) evaluates to 0 if x = 0, and 1/x otherwise.
+
+        // Assign α = inv0(x_q - x_p)
+        region.assign_advice(
+            || "α",
+            self.alpha,
+            offset,
+            || {
+                let x_p = x_p.ok_or(Error::SynthesisError)?;
+                let x_q = x_q.ok_or(Error::SynthesisError)?;
+
+                let alpha = if x_q != x_p {
+                    (x_q - x_p).invert().unwrap()
+                } else {
+                    C::Base::zero()
+                };
+                Ok(alpha)
+            },
+        )?;
+
+        // Assign β = inv0(x_p)
+        region.assign_advice(
+            || "β",
+            self.beta,
+            offset,
+            || {
+                let x_p = x_p.ok_or(Error::SynthesisError)?;
+
+                let beta = if x_p != C::Base::zero() {
+                    x_p.invert().unwrap()
+                } else {
+                    C::Base::zero()
+                };
+                Ok(beta)
+            },
+        )?;
+
+        // Assign γ = inv0(x_q)
+        region.assign_advice(
+            || "γ",
+            self.gamma,
+            offset,
+            || {
+                let x_q = x_q.ok_or(Error::SynthesisError)?;
+
+                let gamma = if x_q != C::Base::zero() {
+                    x_q.invert().unwrap()
+                } else {
+                    C::Base::zero()
+                };
+                Ok(gamma)
+            },
+        )?;
+
+        // Assign δ = inv0(y_q + y_p) if x_q = x_p, 0 otherwise
+        region.assign_advice(
+            || "δ",
+            self.delta,
+            offset,
+            || {
+                let x_p = x_p.ok_or(Error::SynthesisError)?;
+                let x_q = x_q.ok_or(Error::SynthesisError)?;
+                let y_p = y_p.ok_or(Error::SynthesisError)?;
+                let y_q = y_q.ok_or(Error::SynthesisError)?;
+
+                let delta = if x_q == x_p {
+                    if y_q != -y_p {
+                        (y_q + y_p).invert().unwrap()
+                    } else {
+                        C::Base::zero()
+                    }
+                } else {
+                    C::Base::zero()
+                };
```

Comment:
My understanding is that `delta` should be equal to `inv0(y_q + y_p)` when `x_q == x_p` regardless of whether or not `y_q != -y_p`. The code as written is the same behavior (it sorta checks if they're additive inverses of each other which means their sum is zero) but to avoid confusion let's be explicit.

Let's actually bring out `inv0()` as a function elsewhere in the code (this same module is fine) that we call from all the spots we need to use it.